File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/StreamingRewriter.java
package com.codepoetics.raffia.streaming.rewriting;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.operations.Updater;
import com.codepoetics.raffia.paths.Path;
import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.visitors.Visitors;
import com.codepoetics.raffia.writers.BasketWriter;

public abstract class StreamingRewriter<T extends BasketWriter<T>> implements FilteringWriter<T> {

  protected T target;
  protected final FilteringWriter<T> parent;
  protected final Updater updater;

  protected StreamingRewriter(T target, FilteringWriter<T> parent, Updater updater) {
    this.target = target;
    this.parent = parent;
    this.updater = updater;
  }

  public static <T extends BasketWriter<T>> FilteringWriter<T> start(T target, Path path, Updater updater) {
    if (path.isEmpty()) {
      return new MatchedRewriter<>(target, null, updater);
    }

    return new StructStartSeekingRewriter<>(target, path, updater);
  }

  private static Updater makeConditionalUpdater(Path path, Updater updater) {
    return path.head().createItemUpdater(path.tail(), updater);
  }

  public static <T extends BasketWriter<T>> FilteringWriter<T> startArray(T target, Path path, FilteringWriter<T> parent, Updater updater) {
    if (path.isEmpty()) {
      return new MatchedRewriter<>(target.beginArray(), parent, updater);
    }

    if (path.head().isConditional()) {
      return new PredicateMatchingRewriter<>(target.beginArray(), parent, makeConditionalUpdater(path, updater));
    }

    return IndexSeekingRewriter.seekingArrayIndex(target.beginArray(), path, parent, updater);
  }

  public static <T extends BasketWriter<T>> FilteringWriter<T> startObject(T target, Path path, FilteringWriter<T> parent, Updater updater) {
    if (path.isEmpty()) {
      return new MatchedRewriter<>(target.beginObject(), parent, updater);
    }

    if (path.head().isConditional()) {
      return new PredicateMatchingRewriter<>(target.beginObject(), parent, makeConditionalUpdater(path, updater));
    }

    return IndexSeekingRewriter.seekingObjectKey(target.beginObject(), path, parent, updater);
  }

  @Override
  public FilteringWriter<T> advance(T newTarget) {
    target = newTarget;
    return this;
  }

  @Override
  public T complete() {
    return target;
  }

  protected FilteringWriter<T> updated(Basket basket) {
    return advance(updater.update(basket).visit(Visitors.writingTo(target)));
  }

}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/StreamingRewriter.kt
package com.codepoetics.raffia.streaming.rewriting

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.functions.Updater
import com.codepoetics.raffia.paths.Path
import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.writers.BasketWriter
import com.codepoetics.raffia.writers.writeTo

abstract class StreamingRewriter<T : BasketWriter<T>> protected constructor(protected var target: T, protected val parent: FilteringWriter<T>?, protected val updater: Updater?) : FilteringWriter<T> {

    override fun advance(newTarget: T): FilteringWriter<T> {
        target = newTarget
        return this
    }

    override fun complete(): T = target

    protected fun updated(basket: Basket): FilteringWriter<T> = advance(updater!!.update(basket).writeTo(target))

    companion object {

        fun <T : BasketWriter<T>> start(target: T, path: Path, updater: Updater): FilteringWriter<T> =
                if (path.isEmpty) MatchedRewriter<T>(target, null, updater)
                else StructStartSeekingRewriter(target, path, updater)

        private fun makeConditionalUpdater(path: Path, updater: Updater): Updater = path.head().createItemUpdater(path.tail(), updater)

        fun <T : BasketWriter<T>> startArray(target: T, path: Path, parent: FilteringWriter<T>, updater: Updater): FilteringWriter<T> =
                if (path.isEmpty) MatchedRewriter<T>(target.beginArray(), parent, updater)
                else if (path.head().isConditional) PredicateMatchingRewriter(target.beginArray(), parent, makeConditionalUpdater(path, updater))
                else IndexSeekingRewriter.seekingArrayIndex(target.beginArray(), path, parent, updater)

        fun <T : BasketWriter<T>> startObject(target: T, path: Path, parent: FilteringWriter<T>, updater: Updater): FilteringWriter<T> =
                if (path.isEmpty) MatchedRewriter(target.beginObject(), parent, updater)
                else if (path.head().isConditional) PredicateMatchingRewriter(target.beginObject(), parent, makeConditionalUpdater(path, updater))
                else IndexSeekingRewriter.seekingObjectKey(target.beginObject(), path, parent, updater)
    }

}
